### PR TITLE
Improve performance of our javascript for fun and profit

### LINKED ISSUE #127: Add option to turn off ajax call
Research adding a setting to turn off ajax. 

ref: https://secure.helpscout.net/conversation/638064316/258495?folderId=956509
https://wordpress.org/support/topic/disable-and-or-reduce-usage-via-ajax-call-ck_add_user_visit/

### PR DESCRIPTION
**Pull request by tnorthcutt**
Closes #127 

## Summary
Before this PR, our javascript makes an ajax request on _every page load_ 😱 

This PR adds a few things:
- An option to turn off loading of our javascript entirely. This is the Nuclear Option™ and will prevent tagging and custom content (using our shortcode, and in the future a Gutenberg block) from functioning.
- Only sends an ajax request to tag subscribers on page load _if_ we are on a page with tags AND we have a subscriber id (either via cookie or URL query param).

## QA
- Use this version of the plugin: https://github.com/ConvertKit/ConvertKit-WordPress/archive/issue/127.zip
- Set up a page or post with a tag assigned to it
- Set up a page or post without a tag assigned to it
- Visit both pages/posts without a `ck_subscriber_id` (e.g. in an incognito window) and verify that no requests are made to `admin-ajax.php`
- Visit a page with a tag assigned to it, with a `ck_subscriber_id` (e.g. by appending `?ck_subscriber_id={id_here}` to the URL) that belongs to the account the site is connected to, and verify that a request is made to `admin-ajax.php`
- Verify that the above page load also triggers the assigned tag being added to the subscriber in ConvertKit
- Visit a page without a tag assigned, but with a `ck_subscriber_id` (e.g. in the URL as above), and verify that no request is made to `admin-ajax.php`
- Check the box in the settings area to disable ConvertKit javascript entirely, then visit pages both with and without tags assigned and verify that `wp-convertkit.js` is not loaded.
- Re-enable ConvertKit javascript in the settings, then visit a page with a ConvertKit form on it, and subscribe through the form. Verify that a request is made to `admin-ajax.php` with `action: ck_get_subscriber` in the Form Data portion of the request.

## Ready for review?
- [x] Add "ready for review" label
- [x] Assign a reviewer (or two)
- [x] post RFR in #wordpress

### CODE REVIEW

File: admin/section/class-convertkit-settings-general.php [hunk 1]
```diff
@@ -215,14 +223,32 @@ public function debug_callback() {
...
 
+	/**
+	 * Renders the input for no_scripts setting
+	 */
+	public function no_scripts_callback() {
+
+		$no_scripts = '';
+		if ( isset( $this->options['no_scripts'] ) && 'on' === $this->options['no_scripts'] ) {
+			$no_scripts = 'checked';
+		}
```
Comment (pothibo) [inline admin/section/class-convertkit-settings-general.php]: Assigning `$no_scripts = ''` and checking for isset() in the condition below seems repetitive, is that the PHP way?

File: resources/frontend/wp-convertkit.js [hunk 2]
```diff
@@ -1,34 +1,38 @@
 jQuery(document).ready(function($) {
 
     // Manage visit cookie
-    var subscriber_id = $.cookie( 'ck_subscriber_id' );
+    var subscriber_id = parseInt($.cookie('ck_subscriber_id')) ? parseInt($.cookie('ck_subscriber_id')) : false;
```
Comment (pothibo) [inline resources/frontend/wp-convertkit.js]: Not a fan of ternary operator, but if you like it, I'm OK with it. it would be nice to have some guideline as to when to use it.

File: resources/frontend/wp-convertkit.js [hunk 3]
```diff
@@ -1,34 +1,38 @@
...
 
-    }).fail(function (response) {
-        if ( window.console && window.console.log ) {
-            console.log( "AJAX ERROR" + response );
-        }
-    });
+        }).fail(function (response) {
+            if ( window.console && window.console.log ) {
+                console.log( "AJAX ERROR" + response );
+            }
```
Comment (pothibo) [inline resources/frontend/wp-convertkit.js]: Should that only be in development?

File: admin/section/class-convertkit-settings-general.php (hunk 1 above)
Comment (tnorthcutt) [reply admin/section/class-convertkit-settings-general.php]: In this case, it's because we either want `$no_scripts` to be an empty string (`''`), or the string `checked`. That's because it's going in the html to check the box (or not) based on the current value from the database.

The `isset()` isn't checking `$no_scripts`; it's checking the `no_scripts` key in the `$this->options` array.

Make sense?

File: resources/frontend/wp-convertkit.js (hunk 2 above)
Comment (tnorthcutt) [reply resources/frontend/wp-convertkit.js]: @pothibo I am pretty partial to it, but that's fairly subjective/personal.

What would make sense to you as a guideline? Anything you can point me to (just guidelines in general, not ternary-specific) for the ConvertKit app I can look at?

File: admin/section/class-convertkit-settings-general.php (hunk 1 above)
Comment (pothibo) [reply admin/section/class-convertkit-settings-general.php]: Yeah, I understood that. It just felt weird looking at it. In ruby we'd do something like this:

```
$no_script = $this->options.fetch("no_scripts", "") == "on"
```
^ pseudo code above

And then we would use the boolean value.

File: resources/frontend/wp-convertkit.js (hunk 3 above)
Comment (tnorthcutt) [reply resources/frontend/wp-convertkit.js]: Good question. It's been there this whole time, so for now I've left it in. I _think_ it could be somewhat useful in debugging a customer site if the tagging function is failing for them, since we'd be able to see that output ourselves when testing, or instruct them on how to look (ok, not likely).

My hunch is it was put there originally as a "well, if there's a failure we should do _something_".

File: resources/frontend/wp-convertkit.js (hunk 2 above)
Comment (pothibo) [reply resources/frontend/wp-convertkit.js]: Nothing really, I think it was meant more as a general discussion/direction as to the guidelines to writing code in the Wordpress plugin.

We have people on either side of the fence on the JS/Ruby side so it's still up for debate. We are , however, starting to use more Rubocop (linter) to set the guideline there. I think my comment was more of a food for thoughts related to how we want this project to look in the future.

Like when we get to the documentation/linter aspect of it, we should think about stuff like ternary operator and make a decision on it. (No embedded ternary, no ternary at all, etc.)

File: admin/section/class-convertkit-settings-general.php (hunk 1 above)
Comment (tnorthcutt) [reply admin/section/class-convertkit-settings-general.php]: Yeah that makes sense. Working on some improvements to this 👏

**Comment by kostiamel** [comment]
All steps are passed!
